File: rust/src/surface.rs
pub trait Surface {
    fn hit(&self, r: ray3) -> bool;
}

pub struct Sphere {
    position: vec3;
    origin: f32;
}

impl Surface for Sphere {
    fn hit(&self, r: ray3) -> bool {
        // let oc: vec3 = vec3::sub
        false
    }
}

File: go/surface.go
package main

// SurfaceType represents an enum for determining the type of a Surface.
type SurfaceType int

const (
	// Sphere represents a sphere
	Sphere SurfaceType = 0
	// Plane represents a plane
	Plane SurfaceType = 1
)

// Surface represents any surface, and contains the relevant variables for any kind of surface. The used attributes of the struct are determined by Type.
type Surface struct {
	Type SurfaceType

	SphereOrigin Vec3
	SphereRadius float64
}
